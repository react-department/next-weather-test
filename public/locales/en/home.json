{
  "search": "Search",
  "wind": "Wind",
  "precip": "Precip",
  "presure": "Presure",
  "humidity": "Humidity",
  "cloud": "Cloud",
  "temp": "Temp",
  "feelsLike": "Feels like"
}
