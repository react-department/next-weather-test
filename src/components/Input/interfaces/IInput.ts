import { ReactNode, InputHTMLAttributes } from 'react';

export interface IInput extends InputHTMLAttributes<HTMLInputElement> {
  elementRight?: ReactNode,
}
