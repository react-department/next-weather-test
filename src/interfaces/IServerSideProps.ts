export default interface IServerSideProps {
  locale: string,
  res: {
    statusCode: number
  }
}
