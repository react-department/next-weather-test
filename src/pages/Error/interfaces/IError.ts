export default interface Ierror {
  statusCode: React.ReactNode;
}
