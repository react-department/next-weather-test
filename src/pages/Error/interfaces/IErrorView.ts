export default interface IErrorView {
  statusCode: React.ReactNode;
}
