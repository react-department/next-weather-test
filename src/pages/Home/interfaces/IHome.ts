import { TCity } from './TCity';
import { IWeather } from './IWeather';

export default interface IHome {
  search: string;
  setSearch: (value: string) => void;
  foundCities: TCity[];
  onSelectCity: (city: TCity) => void;
  weatherHistory: IWeather[]
  weather?: IWeather
}
