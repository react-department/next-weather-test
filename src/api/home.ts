import axios from '../services/axios';
import { TCity } from '../pages/Home/interfaces/TCity';
import { IWeather } from '../pages/Home/interfaces/IWeather';

export const getCities = async (searchValue: string) => {
  const { data } = await axios.get<TCity[]>(`/search.json?q=${searchValue}`);
  return data;
};
export const getWeather = async (cityId: number) => {
  const { data } = await axios.get<IWeather>(`/current.json?q=id:${cityId}`);
  return data;
};

export const getSearchHistoryWeather = async (cityIds: number[]) => {
  const requests = cityIds.map((cityId) => (
    axios.get<IWeather>(`/current.json?q=id:${cityId}`)
  ));

  const responses = await Promise.all(requests);
  return responses.map((response) => response.data);
};
