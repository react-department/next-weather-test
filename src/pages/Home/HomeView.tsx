import React, { FC } from 'react';
import Layout from '../../layout/Layout';
import Search from './components/Search';
import IHome from './interfaces/IHome';
import WeatherItem from './components/WeatherItem';

const HomeView: FC<IHome> = ({
  search, setSearch, foundCities, onSelectCity, weather, weatherHistory,
}) => (
  <Layout>
    <div className="flex justify-center items-center">
      <Search search={search} setSearch={setSearch} foundCities={foundCities} onSelectCity={onSelectCity} />
    </div>
    <div className="flex p-3 -mx-3 flex-wrap">
      {weather && (
      <>
        <h2 className="text-4xl font-bold dark:text-white w-full">
          {`${weather?.location.name} ${weather.location.region}`}
        </h2>
        <WeatherItem weather={weather} />
      </>
      )}
    </div>
    <div className="flex p-3 -mx-3 flex-wrap">
      {weatherHistory && (
      <>
        <h2 className="text-4xl font-bold dark:text-white w-full">History</h2>
        <hr className="w-full my-3" />
        {weatherHistory.map((item) => (
          <>
            <h3 className="text-2xl font-bold dark:text-white w-full mt-6">
              {`${item?.location.name} ${item.location.region}`}
            </h3>
            <WeatherItem key={item.location.name} weather={item} />
          </>
        ))}
      </>
      )}
    </div>
  </Layout>
);

export default HomeView;
