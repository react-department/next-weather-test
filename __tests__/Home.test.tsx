import React from 'react';
import { fireEvent, render, screen } from '@testing-library/react';
import renderer from 'react-test-renderer';
import Home from '../pages/index';
import { TCity } from '../src/pages/Home/interfaces/TCity';
import HomeView from '../src/pages/Home/HomeView';
import { getCities } from '../src/api/home';

describe('Home', () => {
  it('renders Search input', () => {
    render(<Home />);
    const searchInput = screen.getByRole('searchbox');
    expect(searchInput).toBeInTheDocument();
  });
  it('renders homepage unchanged', () => {
    const tree = renderer.create(<Home />).toJSON();
    expect(tree).toMatchSnapshot();
  });
  it('adds item to localStorage on click', () => {
    localStorage.clear();

    const cities: TCity[] = [{
      id: 1,
      name: 'London',
      region: 'City of London',
      country: 'United Kingdom',
      lat: 51.52,
      lon: -0.11,
      url: 'test',
    }];

    React.useState = jest.fn().mockReturnValue([cities, {}]);

    render(<HomeView
      foundCities={cities}
      onSelectCity={() => localStorage.setItem('historySearch', JSON.stringify([cities[0].id]))}
      search={cities[0].name}
      setSearch={() => {}}
      weather={undefined}
      weatherHistory={[]}
    />);

    const listItem = screen.getByRole('button', { name: /London/i });
    fireEvent.click(listItem);

    const storedHistory = JSON.parse(localStorage.getItem('historySearch') as string) as number[];

    expect(storedHistory).toContain(cities[0].id);
  });
});

describe('getCities', () => {
  describe('when API call is successful', () => {
    it('should return cities list', async () => {
      const cities = [
        {
          id: 2801268,
          name: 'London',
          region: 'City of London, Greater London',
          country: 'United Kingdom',
          lat: 51.52,
          lon: -0.11,
          url: 'london-city-of-london-greater-london-united-kingdom',
        },
        {
          id: 315398,
          name: 'London',
          region: 'Ontario',
          country: 'Canada',
          lat: 42.98,
          lon: -81.25,
          url: 'london-ontario-canada',
        },
      ];
      const result = await getCities('london');

      expect(result[0]).toMatchObject(cities[0]);
      expect(result).toEqual(cities);
    });
  });
});
