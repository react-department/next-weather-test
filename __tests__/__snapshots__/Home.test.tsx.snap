// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home renders homepage unchanged 1`] = `
<div
  className="container"
>
  <main
    className="main"
  >
    <div
      className="flex justify-center items-center"
    >
      <div
        className="w-full"
      >
        <div
          className="relative"
        >
          <div
            className="absolute inset-y-0 start-0 flex items-center ps-3 pointer-events-none"
          >
            <img
              alt="search"
              aria-hidden="true"
              className="w-5 h-5 text-gray-400"
              src="/search.svg"
            />
          </div>
          <div>
            <div>
              <input
                className="block w-full p-4 ps-10 text-sm text-gray-900 border border-gray-300 rounded-lg bg-gray-50 focus:ring-blue-500 focus:border-blue-500 dark:bg-gray-700 dark:border-gray-600 dark:placeholder-gray-400 dark:text-white dark:focus:ring-blue-500 dark:focus:border-blue-500"
                disabled={false}
                onBlur={[Function]}
                onBlurCapture={[Function]}
                onChange={[Function]}
                onFocus={[Function]}
                onFocusCapture={[Function]}
                placeholder="search"
                type="search"
                value=""
              />
            </div>
            <div />
          </div>
        </div>
      </div>
    </div>
    <div
      className="flex p-3 -mx-3 flex-wrap"
    />
    <div
      className="flex p-3 -mx-3 flex-wrap"
    >
      <h2
        className="text-4xl font-bold dark:text-white w-full"
      >
        History
      </h2>
      <hr
        className="w-full my-3"
      />
    </div>
  </main>
  <footer
    className="footer"
  >
    <a
      href="https://onix-systems.com"
      rel="noopener noreferrer"
      target="_blank"
    >
      poweredBy
       
      onix
    </a>
  </footer>
</div>
`;
